temporal-sdk/src/main/java/io/temporal/common/converter/DataConverter.java

```diff
@@ -108,6 +110,28 @@ <T> T fromPayloads(
       int index, Optional<Payloads> content, Class<T> valueType, Type valueGenericType)
       throws DataConverterException;
 
+  /**
+   * Instantiate an appropriate Java Exception from a serialized Failure object. The default
+   * implementation delegates the conversion process to an instance of {@link
+   * io.temporal.failure.FailureConverter}, using this data converter for payload decoding.
+   *
+   * @param failure Failure protobuf object to deserialize into an exception
+   * @throws NullPointerException if failure is null
+   */
+  @Nonnull
+  RuntimeException failureToException(@Nonnull Failure failure);
```

Comment:
The return value can and should be changed to a more specific TemporalFailure, will give much greater peace of mind for the calling code.

Reply:
I think _technically_ we're allowing users to convert these to/from any exception they want

Reply:
It doesn't make much sense. Any exception other than TemporalFailure in workflow code will cause a workflow task failure, while this thing is persisted in the history already. I don't see any legitimate usecase for non-TemporalFailure to be here.
A user that will try to deserialize a Failure into non-TemporalFailure will face issues down the line.

Reply:
I agree. Other exceptions should be wrapped into some `TemporalFailure`, so that makes sense as the return value from both `DataConverter.failureToException` and `FailureConverter.failureToException`.